Add new model docs (#9667)
huggingface/transformers: docs/source/add_new_model.md

> # How to add a model to 🤗Transformers

Reviewer: Nit: I think it would be best to align with other pages and use 🤗 Transformers (with a space) everywhere.

Adding a new model is often difficult and requires an in-depth knowledge of the 

> 🤗Transformers repository and ideally also of the model's original repository.

Reviewer: ```suggestion
🤗Transformers library and ideally also of the model's original repository.
```
(just to avoid repository twice).

At Hugging Face, we are trying to empower the community more and more to add models 
independently. Thus, for some new models that the community wants to be added 
to 🤗Transformers, we create a customized *call-for-model-addition* that explains 
step-by-step how to add the requested model. 
With this *call-for-model-addition*, we want to teach a motivated and experienced 

> contributor of the community how to port a model to Transformers.

Reviewer: ```suggestion
contributor of the community how to port a model to 🤗 Transformers.
```

If this sounds like something you would be interested in, feel free to check out 
the currently open "calls-for-model-addition" [here](https://github.com/huggingface/transformers/tree/master/templates/adding_a_new_model/open_model_proposals/README.md) and to contact us.

If selected, you will then work closely with one member of the Hugging Face team to
integrate the model into 🤗Transformers. By doing so, you will both gain a 
theoretical and deep practical understanding of the proposed model. But more importantly, 
you will have made a major open-source contribution to 🤗Transformers. Along the way,
you will:
	
- get insights into open-source best practices, 
- understand the design principles of one of the most popular NLP libraries,
- learn how to do efficiently test large NLP models,
- learn how to integrate python utilities like `black`, `isort`, `make fix-copies` into a library 

>   to always ensure clean and readable code.

Reviewer: No mention of certificate here?

We are obviously also more than happy if you want to add a model which can not be found in 
the "calls-for-model-addition" folder. The following sections explain in detail how to add a new model. It might also be very helpful 
to check-out already added models to see if those resemble the model you would like 
to add [here](https://github.com/huggingface/transformers/pulls?q=is%3Apr+label%3A%22PR+for+Model+Addition%22+is%3Aclosed).

To start with let's try to get a general overview of the Transformers library.

## General overview of 🤗Transformers

TODO

- Mention some general design principles
- Mention the single file policy
- Mention the copy over abstraction policy
- ...

### Overview of models

TODO

- Add graphic

### Overview of tokenizers

TODO

- Add graphic

## Adding brand_new_bert to 🤗Transformers

> For exemplary purposes, we will call the model that we will add to 🤗Transformers `"brand_new_bert"`.

Reviewer: I prefered LysBert :-p

  
The following list is a summary of everything that has to be done to add a model and can be used 
by you as a To-Do List:

- [ ] Understood theoretical aspects
- [ ] Prepared environment
- [ ] Set up debugging environment of original repository
- [ ] Created script that successfully runs forward pass using original repository and checkpoint

> - [ ] Successfully converted original checkpoint to Transformers checkpoint

Reviewer: I would add - [] Successfully added Model in Transformers before this conversion script.

- [ ] Successfully ran forward pass in Transformers that gives identical output to original checkpoint
- [ ] Finished model tests in Transformers
- [ ] Succesfully added Tokenizer in Transformers
- [ ] Finished docs
- [ ] Uploaded model weights to the hub
- [ ] Merged the pull request

> - [ ] (Optionally) added a notebook

Reviewer: Great TODO!

To begin with, you should start by getting a good understanding of the model.
  
### Theoretical aspects of brand_new_bert
  
#### Paper

  You should take some time to read brand_new_bert's paper.
  There might be large sections of the paper that are difficult to understand.
  If this is the case, this is fine - don't worry! The goal is not to get 
  a deep theoretical understanding of the paper, but to extract the necessary information 
  required to effectively re-implement the model to 🤗Transformers.
  That being said, you don't have to spend too much time on the theoretical aspects,
  but rather focus on the practical ones, namely:
  
  - What time of model is brand_new_bert? BERT-like encoder-only model? GPT2-like decoder-only model? BART-like encoder-decoder model?
  - What are the applications of brand_new_bert? Text classification? Text generation? Seq2Seq tasks, *e.g.* summarization?
  - What is the novel feature of the model making it different from BERT or BART (if it's an encoder-decoder model)?
  - Which of the already existing [🤗Transformers models](https://huggingface.co/transformers/#contents) is most similar to brand_new_bert?

>   - What type of tokenizer is used? A sentence piece tokenizer? Word piece tokenizer? Is it the same tokenizer as used for BERT or BART?

Reviewer: ```suggestion
  - What type of tokenizer is used? A sentencepiece tokenizer? Word-piece tokenizer? Is it the same tokenizer as used for BERT or BART?
```
Never seen sentencepeice in two words but I might be wrong.

  
 After you feel like you have gotten a good overview of the architecture of the model, you might want 
 to write [name of hugging face teacher] for any questions you might have.
 This might include questions regarding the model's architecture, its attention layer, etc. [name of hugging face teacher] will be more 
 than happy to help you.
 
### Next prepare your environment

1. Fork the [repository](https://github.com/huggingface/transformers) by clicking on the 'Fork' button on the repository's page.
This creates a copy of the code under your GitHub user account.

2. Clone your fork to your local disk, and add the base repository as a remote:

	```bash
	git clone https://github.com/[your Github handle]/transformers.git
	cd transformers
	git remote add upstream https://github.com/huggingface/transformers.git
	```
 
3. Set up a development environment, for instance by running the following command:

	```bash
	conda create -n env python=3.7 --y
	conda activate env
	pip install -e ".[dev]"
	```
  
  and return to the parent directory
  
  ```bash
  cd ..
  ```

4. We recommend to add the PyTorch version of brand_new_bert to Transformers. To install PyTorch,

>   please follow the instructions on https://pytorch.org/. 

Author: Agree - I'll leave that for a v2. All the code for conversion is only written in PyTorch, etc...

  
  **Note:** You don't need to have CUDA install. It is sufficient to just be working on CPU.

5. To port brand_new_bert, you will also need access to its original repository:
  
  ```bash
  git clone https://github.com/brand_new_bert_org/brand_new_bert.git 
  cd brand_new_bert
  pip install -e .
  ```

Now you have set up a development environment to port brand_new_bert to 🤗Transformers.
  
### Run a pretrained checkpoint using the original repository

At first, you will work on the original repository. Often, the original implementation is very
 "researchy" meaning that documentation might be lacking and the code can be difficult to understand.
But this should be exactly your motivation to reimplement brand_new_bert. At Hugging Face, one of our main goals is to
*make people stand on the shoulders of giants* which translates here very well into taking a working 
model and rewriting it to make it as **accessible, user-friendly, and beautiful** as possible.

> This is the #1 motivation to re-implement models into 🤗Transformers - trying to maximize access 

Reviewer: ```suggestion
This is the number-one motivation to re-implement models into 🤗Transformers - trying to maximize access 
```

to a complex new NLP technology for **everybody**.
	
You should start thereby by diving into the original repository.

> #### Get familiar with the original repository.

Reviewer: I am not sure where to mention this, but everything in this step can be (and should be) done on a CPU. We should clearly state that there is no need to setup a costly env with GPU before some integration training test at the very end.

> Successfully running the official pre-trained model in the original repository is often 

Reviewer: I'm very partial to pretrained so would recommend all pre-trained -> pretrained.

**the most difficult** step. From our experience, it is very important to spend some time to get familiar with the original codebase. You should find out

> - Where to find the pre-trained weights
> - How to load the pre-trained weights into its corresponding model
> - Trace one forward pass so that you know which classes and functions are required for a simple forward pass
>   . Usually, you only have to reimplement those functions.
> - Be able to locate the important components of the model: Where is the model class? Are there submodel 
>   classes, *e.g.* EncoderModel, DecoderModel? Where is the self-attention layer? 
>   Are there multiple different attention layers, *e.g.* *self-attention*, *cross-attention*...?
> - How can you debug the model in the original environment of the repo? Do you have to set `print` statements
>   or can you work with an interactive debugger like `ipdb`?

Reviewer: ```suggestion
- Where to find the pre-trained weights?
- How to load the pre-trained weights into its corresponding model?
- Trace one forward pass so that you know which classes and functions are required for a simple forward pass
  . Usually, you only have to reimplement those functions.
- Be able to locate the important components of the model: Where is the model class? Are there submodel 
  classes, *e.g.* EncoderModel, DecoderModel? Where is the self-attention layer? 
  Are there multiple different attention layers, *e.g.* *self-attention*, *cross-attention*...?
- How can you debug the model in the original environment of the repo? Do you have to set `print` statements
  or can you work with an interactive debugger like `ipdb`?
```

  
It is very important that before you start opening a PR in 🤗Transformers that you can **efficiently** 

> debug code in the original repository! This means that you should be able to run a forward pass and print out the actual values of the output of a layer. *I.e* you can load a pre-trained model, pass an input vector of 

Reviewer: ```suggestion
debug code in the original repository! This means that you should be able to run a forward pass and print out the actual values of the output of a layer. Make sure you can load a pre-trained model, pass an input vector of 
```

token ids, *i.e.* `input_ids = [0, 1, 4, 5, ...]` to the model's forward function and you can print out the 
intermediate outputs of - let's say - the first self-attention layer that could look something like this: 

```bash
[[
 [-0.1465, -0.6501,  0.1993,  ...,  0.1451,  0.3430,  0.6024],
 [-0.4417, -0.5920,  0.3450,  ..., -0.3062,  0.6182,  0.7132],
 [-0.5009, -0.7122,  0.4548,  ..., -0.3662,  0.6091,  0.7648],
 ...,
 [-0.5613, -0.6332,  0.4324,  ..., -0.3792,  0.7372,  0.9288],
 [-0.5416, -0.6345,  0.4180,  ..., -0.3564,  0.6992,  0.9191],
 [-0.5334, -0.6403,  0.4271,  ..., -0.3339,  0.6533,  0.8694]]],

> ```

Reviewer 2: While printing can get things done, I think there are better ways to go about it, which may be a bit more complex to setup, but end up being **so** useful as development goes on.

When porting a model, I like to:
- Isolate the different forward passes that are useful, in a step-by-step manner. This means that I like to proceed as follows:
    - Have a way to retrieve the embeddings
    - Have a way to retrieve each layer's output
    - Have a way to retrieve the Transformer output
    - Have a way to retrieve the full model output, for each head
- All of these output should be gathered into a single script, in **variables**. This can be a pain to setup, but makes it so much simpler in the end because you can **compare** values.
- These values should be compared to the implementation in HuggingFace Transformers. A relative/absolute difference can be defined

Having this step-by-step approach with the ability to compare is imo incredibly useful because:
- You can check the results algorithmically, rather than checking the terminal output and verifying the results
- You can set this up very early on, starting with the embeddings
- You can ensure you have no regression when working on the model, as when finishing up the last head, you'll still be checking the embeddings. You'll be checking all heads at once, too.

Here's an example of how I did it with ELECTRA: [compare_electra.py](https://gist.github.com/LysandreJik/db4c948f6b4483960de5cbac598ad4ed). It's slightly more complicated to setup but I think it's worthwhile.

I have no idea how you would explain this in the document, so feel free to ignore if it's not appropriate :)

Author: Yes, I think that's the more in-depth approach! I did this for Reformer as well. Maybe write down both approaches?

Reviewer: I personally prefer to have everything in a notebook where I develop new components and import old components to check the outputs of every single object match. I think we should also show that approach, with some "choose your own adventure" question: are you more comfortable with notebooks or scripts? In case a..., in case b...

This means that your debugging environment should consists of a short script (ideally written by you) that 
does the following (in pseudocode):

```bash
model = brand_new_bertModel.load_pretrained_checkpoint(/path/to/checkpoint/)
input_ids = ... # vector of input ids
original_output = model.predict(input_ids)
```

By running such a script, you should be able to print out intermediate values or hit a breakpoint
in the clone of the original repository that is saved locally on your computer.

We expect that every model added to 🤗Transformers passes a couple of integration tests, meaning that the original 

> model and the reimplemented version in 🤗Transformers have to give the exact same output up to a precision of 0.001! 

Reviewer 2: I would explain why "0.001" too! Not necessarily the value you chose to showcase, but explaining that some difference can happen between frameworks and that's okay.

  are available, it might make more sense to create a dummy model in the new environment with randomly initialized weights 
  and save those weights for comparison with the 🤗Transformers version of your model
- Make sure you are using the easiest way of calling a forward pass in the original repository. Ideally, you want to find the function
  in the original repository that **only** calls a single forward pass, *i.e.* that is often called `predict`, `evaluate`, `forward` or `__call__`. 
  You don't want to debug a function that calls `forward` multiple times, *e.g.* to generate text, like `autoregressive_sample`, `generate`.
- Try to separate the tokenization from the model's `forward` pass. If the original repository shows examples where you have to input a string, then try to find out where in the forward call the string input is changed to input ids and start from this point. This might mean that you have to possibly write a small script yourself or change the original code so that you can directly input the ids instead of an input string.

> - Make sure that the model in your debugging setup is **not** in training mode, which often causes the model to yield random outputs due to multiple dropout layers in the model. Make sure that the forward pass in your debugging environment is **deterministic** so that the dropout layers are not used.

Reviewer: Or use `transformers.file_utils.set_seed` if the old and new implementations are in the same framework.

necessary code is added. From our experience, it is much more efficient to quickly add a first version of the required code
and improve/correct the code iteratively with the conversion script as described in the next section. The only thing that has to work at this 
point is that you can instantiate the 🤗Transformers implementation of brand_new_bert, *i.e.* the following command works:
	
```python 
from transformers import BrandNewBertModel, BrandNewBertConfig

model = BrandNewBertModel(BrandNewBertConfig())
```

The above command will create a model according to the default parameters as defined in `BrandNewBertConfig()`
with random weights, thus making sure that the `init()` methods of all components works.

#### Write a conversion script

Now, you should write a conversion script that lets you convert the checkpoint you used to debug 
brand_new_bert in the original repository to your just created 🤗Transformers implementation of brand_new_bert.
Here you should not try to write the conversion script from scratch, but find similar models in 🤗Transformers
that require similar conversion scripts, *i.e.* whose original repository was written with the same framework as

> brand_new_bert.

Reviewer 2: What about the case where a model can be converted to an existing architecture? Like the BORT model, there's no need to go through the last previous steps - just create a conversion script and test that the results are accurate.

Reviewer 2: This whole paragraph is cool - but it's kind of reinventing the wheel. I would add a link to an existing conversion script and advise the user to modify it to fit their needs, rather than write it all from scratch.

Author: Agree, my motivation was not to re-invent the wheel, but rather explaining:

1) PyTorch class attribute names correspond to the checkpoint layer names
2) Why one should also make sure that name and size match

In the end of the paragraph it's probably better to emphasize that one should use an already written conversion script

**Important**: In PyTorch, the name of a layer is defined by the name of the class attribute you give the layer. *E.g.*, let's define a dummy model in PyTorch, called `SimpleModel` as follows:

```python
import torch.nn as nn

class SimpleModel(nn.Module):

    def __init__(self):
		    super().__init__()
				self.dense = nn.Linear(10, 10)
				self.intermediate = nn.Linear(10, 10)

> 				self.layer_norm = nn.LayerNorm(10)

Reviewer: Formatting is weird here.

requires the weight to be transposed beforehand.

Finally, you should also check that **all** required weights are initialized and print
out all checkpoint weights that were not used for initialization to make sure the model is correctly converted.
It is completely normal, that the conversion trials fail with either a wrong shape statement or wrong name assignment.
This is most likely because either you used incorrect parameters in `BrandNewBertConfig()`, have a wrong architecture 
in the 🤗Transformers implementation, you have a bug in the `init()` functions of one of the components of 
the 🤗Transformers implementation or you need to transpose one of the checkpoint weights.

This step should be iterated with the previous step until all weights of the checkpoint are correctly loaded in the Transformers model.
Having correctly loaded the checkpoint into the 🤗Transformers implementation, you can then save the model under a folder of your choice `/path/to/converted/checkpoint/folder` that should include both a `pytorch_model.bin` file and a `config.json` file.

#### Implement the forward pass

> Having managed to correctly load the pretrained weights into the 🤗Transformers implementation, you should now make sure that the forward pass is correctly implemented. In *Get familiar with the original repository.*, you have already created a script

Reviewer 2: This can probably be a link: `Get familiar with the original repository`

that runs a forward pass of the model using the original repository. Now you should write an analogous script using the 
🤗Transformers implementation instead of the original one. It should look as follows:

```python
model = brand_new_bertModel.from_pretrained(/path/to/converted/checkpoint/folder)
input_ids = ... # the exact same vector of input ids in PyTorch as those used in the *Get familiar with the original repository.* section
output = model(input_ids).last_hidden_states
```

> It is very likely that the 🤗Transformers implementation and the original model implementation don't give the exact same output the very first time or that the forward pass throws an error. Don't be disappointed - it's expected! First, you should make sure that the forward pass doesn't throw any errors. It often happens that the wrong dimensions are used leading to a `Dimensionality mismatch` error or that the wrong data type object is used,

Reviewer 2: This part is great! It got me thinking, it could be nice to do a kind of "FAQ", where the headers are the errors that can happen. 

We can fill it up as we encounter them. What do you think? (Potential v2, not necessarily for this version)

*e.g.* `torch.long` instead of `torch.float32`. Don't hesitate to ask the Hugging Face team for help, if you don't manage to solve certain errors.

The final part to make sure the 🤗Transformers implementation works correctly is to ensure that the outputs are equivalent to a precision of `1e-3`.
First, you should ensure that the output shapes are identical, *i.e.* `outputs.shape` should yield the same value for the script of the 🤗Transformers implementation and the original implementation. Next, you should make sure that the output values are identical as well. This one of the most difficult parts of adding a new model. Common mistakes why the outputs are not identical are:

> - Some layers were not added, *i.e.* an `activation` layer was not added, or the residual connection was forgotten
> - The word embedding matrix was not tied
> - The wrong positional embeddings are used because the original implementation uses on offset
> - Dropout is applied during the forward pass. To fix this make sure `model.training is False` and that no dropout layer is falsely activated during the 
>   forward pass, *i.e.* pass `self.training` to [PyTorch's functional dropout](https://pytorch.org/docs/stable/nn.functional.html?highlight=dropout#torch.nn.functional.dropout)

Reviewer: ```suggestion
- Some layers were not added, *i.e.* an `activation` layer was not added, or the residual connection was forgotten.
- The word embedding matrix was not tied.
- The wrong positional embeddings are used because the original implementation uses on offset.
- Dropout is applied during the forward pass. To fix this make sure `model.training is False` and that no dropout layer is falsely activated during the 
  forward pass, *i.e.* pass `self.training` to [PyTorch's functional dropout](https://pytorch.org/docs/stable/nn.functional.html?highlight=dropout#torch.nn.functional.dropout).
```

```python
pytest tests/test_modeling_brand_new_bert.py
```

Having fixed all common tests, it is now crucial to ensure that all the nice work you have done is well tested, so that 

- a) The community can easily understand your work by looking at specific tests of brand_new_bert
- b) Future changes to your model will not break any important feature of the model.

At first, integration tests should be added. Those integration tests essentially do the same as the debugging scripts you used earlier to implement the model to 🤗Transformers. A template of those model tests is already added by the Cookiecutter, called 
`BrandNewBertModelIntegrationTests` and only has to be filled out by you. To ensure that those tests are passing, 
run 

```python
RUN_SLOW=1 pytest tests/test_modeling_brand_new_bert.py::BrandNewBertModelIntegrationTests
```

Second, all features that are special to brand_new_bert should be tested additionally in a separate test under `BrandNewBertModelTester`/`BrandNewBertModelTest`. This part is often forgotten, but is extremely useful in two ways:

> - It helps to transfer the knowledge you have acquired during the model addition to the community by showing how the special feature of brand_new_bert should work
> - Future contributors can quickly test changes to the model by running those special tests

Reviewer: ```suggestion
- It helps to transfer the knowledge you have acquired during the model addition to the community by showing how the special feature of brand_new_bert should work.
- Future contributors can quickly test changes to the model by running those special tests.
```

 
 Analogous to the modeling test files of brand_new_bert, the tokenization test files of brand_new_bert should 
 contain a couple of hard-coded integration tests.
 
 As a final step, it is recommended to add one large integration tests to `tests/test_modeling_brand_new_bert.py` that does some end-to-end
 testing using both the model and the tokenizer.
 
#### Add Docstring

Now, all the necessary functionality for brand_new_bert is added - you're almost done! The only thing left to add is a nice docstring and a doc page. The Cookiecutter should have added a template file called `docs/source/model_doc/brand_new_bert.rst`
that you should fill out. Users of your model will usually first look at this page before using your model. Hence, the documentation must be understandable and concise. It is very useful for the community to add some *Tips* to show how the model should be used.
Don't hesitate to ping the Hugging Face team regarding the docstrings.

Next, make sure that the docstring added to `src/transformers/models/brand_new_bert/modeling_brand_new_bert.py` is correct and included all necessary

> inputs and outputs. It is always to good to remind oneself that documentation should be treated at least as carefully as the code in 🤗Transformers since the documentation is usually the first contact point of the community with the model.

Reviewer 2: Love that!
